Packet++/src/IPv4Layer.cpp

```diff
@@ -261,72 +261,79 @@ void IPv4Layer::parseNextLayer()
 		return;
 	}
 
-	switch (ipHdr->protocol)
+	try
 	{
-	case PACKETPP_IPPROTO_UDP:
-		if (payloadLen >= sizeof(udphdr))
-			m_NextLayer = new UdpLayer(payload, payloadLen, this, m_Packet);
-		break;
-	case PACKETPP_IPPROTO_TCP:
-		m_NextLayer = TcpLayer::isDataValid(payload, payloadLen)
-			? static_cast<Layer*>(new TcpLayer(payload, payloadLen, this, m_Packet))
-			: static_cast<Layer*>(new PayloadLayer(payload, payloadLen, this, m_Packet));
-		break;
-	case PACKETPP_IPPROTO_ICMP:
-		m_NextLayer = IcmpLayer::isDataValid(payload, payloadLen)
-			? static_cast<Layer*>(new IcmpLayer(payload, payloadLen, this, m_Packet))
-			: static_cast<Layer*>(new PayloadLayer(payload, payloadLen, this, m_Packet));
-		break;
-	case PACKETPP_IPPROTO_IPIP:
-		ipVersion = *payload >> 4;
-		if (ipVersion == 4)
-			m_NextLayer = new IPv4Layer(payload, payloadLen, this, m_Packet);
-		else if (ipVersion == 6)
-			m_NextLayer = new IPv6Layer(payload, payloadLen, this, m_Packet);
-		else
-			m_NextLayer = new PayloadLayer(payload, payloadLen, this, m_Packet);
-		break;
-	case PACKETPP_IPPROTO_GRE:
-		greVer = GreLayer::getGREVersion(payload, payloadLen);
-		if (greVer == GREv0)
-			m_NextLayer = new GREv0Layer(payload, payloadLen, this, m_Packet);
-		else if (greVer == GREv1)
-			m_NextLayer = new GREv1Layer(payload, payloadLen, this, m_Packet);
-		else
-			m_NextLayer = new PayloadLayer(payload, payloadLen, this, m_Packet);
-		break;
-	case PACKETPP_IPPROTO_IGMP:
-		igmpVer = IgmpLayer::getIGMPVerFromData(payload, be16toh(getIPv4Header()->totalLength) - hdrLen, igmpQuery);
-		if (igmpVer == IGMPv1)
-			m_NextLayer = new IgmpV1Layer(payload, payloadLen, this, m_Packet);
-		else if (igmpVer == IGMPv2)
-			m_NextLayer = new IgmpV2Layer(payload, payloadLen, this, m_Packet);
-		else if (igmpVer == IGMPv3)
+		switch (ipHdr->protocol)
 		{
-			if (igmpQuery)
-				m_NextLayer = new IgmpV3QueryLayer(payload, payloadLen, this, m_Packet);
+		case PACKETPP_IPPROTO_UDP:
+			if (payloadLen >= sizeof(udphdr))
+				m_NextLayer = new UdpLayer(payload, payloadLen, this, m_Packet);
+			break;
+		case PACKETPP_IPPROTO_TCP:
+			m_NextLayer = TcpLayer::isDataValid(payload, payloadLen)
+				? static_cast<Layer*>(new TcpLayer(payload, payloadLen, this, m_Packet))
+				: static_cast<Layer*>(new PayloadLayer(payload, payloadLen, this, m_Packet));
+			break;
+		case PACKETPP_IPPROTO_ICMP:
+			m_NextLayer = IcmpLayer::isDataValid(payload, payloadLen)
+				? static_cast<Layer*>(new IcmpLayer(payload, payloadLen, this, m_Packet))
+				: static_cast<Layer*>(new PayloadLayer(payload, payloadLen, this, m_Packet));
+			break;
+		case PACKETPP_IPPROTO_IPIP:
+			ipVersion = *payload >> 4;
+			if (ipVersion == 4)
+				m_NextLayer = new IPv4Layer(payload, payloadLen, this, m_Packet);
+			else if (ipVersion == 6)
+				m_NextLayer = new IPv6Layer(payload, payloadLen, this, m_Packet);

```

Comment:
Instead of `try..catch` a better solution would be to check `IPv6Layer::isDataValid(...)` or `IPv4Layer::isDataValid(...)`, like this:
```cpp
ipVersion = *payload >> 4;
if (ipVersion == 4 && IPv4Layer::isDataValid(payload, payloadLen))
	m_NextLayer = new IPv4Layer(payload, payloadLen, this, m_Packet);
else if (ipVersion == 6 && IPv6Layer::isDataValid(payload, payloadLen))
	m_NextLayer = new IPv6Layer(payload, payloadLen, this, m_Packet);
else
	m_NextLayer = new PayloadLayer(payload, payloadLen, this, m_Packet);
```

Reply:
Didn't notice the pattern.